dartagnan/src/main/java/com/dat3m/dartagnan/expression/op/COpBin.java

```diff
@@ -34,28 +36,73 @@ public String toString() {
+            	if(e1 instanceof IntegerFormula) {
+            		IntegerFormulaManager imgr = ctx.getFormulaManager().getIntegerFormulaManager();
+            		return imgr.lessThan((IntegerFormula)e1, (IntegerFormula)e2);
+            	}
+            	if(e1 instanceof BitvectorFormula) {
+            		BitvectorFormulaManager bvmgr = ctx.getFormulaManager().getBitvectorFormulaManager();
+            		return bvmgr.lessThan((BitvectorFormula)e1, (BitvectorFormula)e2, this.equals(LT));
+            	}
             case LTE:
-                return e1.isBV() ? ctx.mkBVSLE((BitVecExpr)e1, (BitVecExpr)e2) : ctx.mkLe((ArithExpr)e1, (ArithExpr)e2);
             case ULTE:
-                return e1.isBV() ? ctx.mkBVULE((BitVecExpr)e1, (BitVecExpr)e2) : ctx.mkLe((ArithExpr)e1, (ArithExpr)e2);
+            	if(e1 instanceof IntegerFormula) {
+            		IntegerFormulaManager imgr = ctx.getFormulaManager().getIntegerFormulaManager();
+            		return imgr.lessOrEquals((IntegerFormula)e1, (IntegerFormula)e2);
+            	}
+            	if(e1 instanceof BitvectorFormula) {
+            		BitvectorFormulaManager bvmgr = ctx.getFormulaManager().getBitvectorFormulaManager();
+            		return bvmgr.lessOrEquals((BitvectorFormula)e1, (BitvectorFormula)e2, this.equals(LTE));
+            	}
             case GT:
-            	return e1.isBV() ? ctx.mkBVSGT((BitVecExpr)e1, (BitVecExpr)e2) : ctx.mkGt((ArithExpr)e1, (ArithExpr)e2);
             case UGT:
-            	return e1.isBV() ? ctx.mkBVUGT((BitVecExpr)e1, (BitVecExpr)e2) : ctx.mkGt((ArithExpr)e1, (ArithExpr)e2);
+            	if(e1 instanceof IntegerFormula) {
+            		IntegerFormulaManager imgr = ctx.getFormulaManager().getIntegerFormulaManager();
+            		return imgr.greaterThan((IntegerFormula)e1, (IntegerFormula)e2);
+            	}
+            	if(e1 instanceof BitvectorFormula) {
+            		BitvectorFormulaManager bvmgr = ctx.getFormulaManager().getBitvectorFormulaManager();
+            		return bvmgr.greaterThan((BitvectorFormula)e1, (BitvectorFormula)e2, this.equals(GT));
+            	}
             case GTE:
-            	return e1.isBV() ? ctx.mkBVSGE((BitVecExpr)e1, (BitVecExpr)e2) : ctx.mkGe((ArithExpr)e1, (ArithExpr)e2);
             case UGTE:
-            	return e1.isBV() ? ctx.mkBVUGE((BitVecExpr)e1, (BitVecExpr)e2) : ctx.mkGe((ArithExpr)e1, (ArithExpr)e2);
+            	if(e1 instanceof IntegerFormula) {
+            		IntegerFormulaManager imgr = ctx.getFormulaManager().getIntegerFormulaManager();
+            		return imgr.greaterOrEquals((IntegerFormula)e1, (IntegerFormula)e2);
+            	}
+            	if(e1 instanceof BitvectorFormula) {
+            		BitvectorFormulaManager bvmgr = ctx.getFormulaManager().getBitvectorFormulaManager();
+            		return bvmgr.greaterOrEquals((BitvectorFormula)e1, (BitvectorFormula)e2, this.equals(GTE));
+            	}
```

Comment:
The checks and the switch should be interchanged, i.e. first check the types of e1 and e2 and then perform a switch on the operation inside each of the branches (bool + bool, int + int, bv + bv). The code in each branch is more clearer because the respective FormulaManager can be stored in a local variable as well as the casted two expressions.